titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/evacuation/EvacuationExecutor.java

```diff
@@ -0,0 +1,215 @@
+                logger.info("Aborting the task termination process due to job quota limits: candidate={}, jobId={}", task.getId(), task.getJobId());
+            }
+        }
+    }
+
+    private void evictTask(Task task) {
+        Throwable error = evictionServiceClient.terminateTask(
+                task.getId(),
+                "Evacuating task from the quarantined agent instance: " + instance.getId()
+        ).get();
+
+        if (error != null) {
+            metrics.evictionFailure(task, error);
+            logger.info("Eviction service task termination error: taskId=={}", task.getId(), error);
+            return;
+        }
+        metrics.evictionSucceeded(task);
+
+        pendingTaskTerminationsByJobId.put(task.getJobId(), task.getId());
+        logger.info("Terminating task: {}", task.getId());
+    }
+
+    private EvictionQuota getCapacityGroupQuota(Job<?> job) {
+        EvictionDataSnapshot evictionData = evictionDataReplicator.getCurrent();
+        return evictionData.
+                findCapacityGroupEvictionQuota(job.getJobDescriptor().getCapacityGroup())
+                .orElseGet(() ->
+                        evictionData
+                                .findCapacityGroupEvictionQuota("DEFAULT")
+                                .orElseGet(() -> EvictionQuota.newBuilder()
+                                        .withReference(Reference.capacityGroup(job.getJobDescriptor().getCapacityGroup()))
+                                        .withQuota(0)
+                                        .build()
+                                )
+                );
+    }
+
+    private boolean hasJobQuota(Job<?> job, Task task) {
+        long desired = JobFunctions.getJobDesiredSize(job);
+        long actual = jobDataReplicator.getCurrent().getTasks(job.getId()).stream()
+                .filter(t -> t.getStatus().getState() == TaskState.Started || t.getId().equals(task.getId()))
+                .count();
+        return desired <= actual;
+    }
+
+    private boolean isTerminated(String taskId) {
+        return jobDataReplicator.getCurrent().findTaskById(taskId)
+                .map(task -> task.getRight().getStatus().getState() == TaskState.Finished)
+                .orElse(true);
+    }
+
+    static List<Task> getTasksRunningOnInstance(JobDataReplicator jobDataReplicator, AgentInstance instance) {
+        return jobDataReplicator.getCurrent().getTasks().stream()
+                .filter(task -> isTaskRunningOnInstance(task, instance.getId()))
+                .collect(Collectors.toList());
+    }
+
+    static boolean isAgentRunningTasks(JobDataReplicator jobDataReplicator, AgentInstance instance) {
+        return jobDataReplicator.getCurrent().getTasks().stream()
+                .anyMatch(task -> isTaskRunningOnInstance(task, instance.getId()));
+    }
+
+    static boolean isTaskRunningOnInstance(Task task, String instanceId) {
+        if (task.getStatus().getState() == TaskState.Finished) {
```

Comment:
Shouldn't you only care about `StartInitiated` and `Started` here?

Reply:
We care here about all task states in which an agent is assigned, except the `Finished` state.

Reply:
We regard tasks in the `KillInitiated` state as running, since they are running on the agent, and the agent cannot be terminated until all of them move to the `Finished` state.